github/downloader.go

```diff
@@ -0,0 +1,921 @@
+package github
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net/http"
+
+	"github.com/src-d/metadata-retrieval/github/graphql"
+	"github.com/src-d/metadata-retrieval/github/store"
+
+	"github.com/shurcooL/githubv4"
+	"gopkg.in/src-d/go-log.v1"
+)
+
+const (
+	assigneesPage                 = 2
+	issueCommentsPage             = 10
+	issuesPage                    = 50
+	labelsPage                    = 2
+	membersWithRolePage           = 100
+	pullRequestReviewCommentsPage = 5
+	pullRequestReviewsPage        = 5
+	pullRequestsPage              = 50
+	repositoryTopicsPage          = 50
+)
+
+type storer interface {
+	SaveOrganization(organization *graphql.Organization) error
+	SaveUser(user *graphql.UserExtended) error
+	SaveRepository(repository *graphql.RepositoryFields, topics []string) error
+	SaveIssue(repositoryOwner, repositoryName string, issue *graphql.Issue, assignees []string, labels []string) error
+	SaveIssueComment(repositoryOwner, repositoryName string, issueNumber int, comment *graphql.IssueComment) error
+	SavePullRequest(repositoryOwner, repositoryName string, pr *graphql.PullRequest, assignees []string, labels []string) error
+	SavePullRequestComment(repositoryOwner, repositoryName string, pullRequestNumber int, comment *graphql.IssueComment) error
+	SavePullRequestReview(repositoryOwner, repositoryName string, pullRequestNumber int, review *graphql.PullRequestReview) error
+	SavePullRequestReviewComment(repositoryOwner, repositoryName string, pullRequestNumber int, pullRequestReviewId int, comment *graphql.PullRequestReviewComment) error
+
+	Begin() error
+	Commit() error
+	Rollback() error
+	Version(v string)
+	SetActiveVersion(v string) error
+	Cleanup(currentVersion string) error
+}
+
+// Downloader fetches GitHub data using the v4 API
+type Downloader struct {
+	storer
+	client *githubv4.Client
+}
+
+// NewDownloader creates a new Downloader that will store the GitHub metadata
+// in the given DB. The HTTP client is expected to have the proper
+// authentication setup
+func NewDownloader(httpClient *http.Client, db *sql.DB) (*Downloader, error) {
+	// TODO: is the ghsync rate limited client needed?
```

Comment:
currently, it exits with an error when the number of requests exceeded the rate limit instead of waiting for reset.

Reply:
Right, this is what I found too. For now I thought it was ok to behave like that, but I opened an issue to have this conversation separately https://github.com/src-d/metadata-retrieval/issues/3

Reply:
it's ok for the lib but not ok for the cli.
It means there is no way to download an organization if it's too big.